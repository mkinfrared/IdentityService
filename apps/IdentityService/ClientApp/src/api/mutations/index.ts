export * from "./loginMutation";

export * from "./registerMutation";

export * from "./verifyEmailMutation";

export * from "./submitConsent";

export * from "./submitForgotPassword";

export * from "./submitPasswordChange";
