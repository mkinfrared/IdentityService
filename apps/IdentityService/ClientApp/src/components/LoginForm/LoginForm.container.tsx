import { memo } from "react";

import { useReturnUrl } from "hooks";

import LoginFormFC from "./LoginForm";
import { LoginFormContainerProps } from "./LoginForm.type";

const LoginFormContainer = (props: LoginFormContainerProps) => {
  const returnUrl = useReturnUrl();

  return <LoginFormFC {...props} returnUrl={returnUrl} />;
};

export default memo(LoginFormContainer);
