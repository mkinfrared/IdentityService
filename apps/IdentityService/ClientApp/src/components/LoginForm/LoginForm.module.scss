@import "../../styles/common";

.LoginForm {
  @extend %login-register-form;

  .forgotPassword {
    text-align: center;
  }
}
