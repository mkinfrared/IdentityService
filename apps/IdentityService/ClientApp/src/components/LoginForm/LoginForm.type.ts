export type LoginFormProps = {
  /**
   * a string that will be applied as a css class to parent element
   */
  className?: string;
  returnUrl: string;
};

export type LoginFormContainerProps = Omit<LoginFormProps, "returnUrl">;
