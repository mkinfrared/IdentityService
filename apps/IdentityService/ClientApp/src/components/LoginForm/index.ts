export { default } from "./LoginForm.container";

export * from "./LoginForm.type";
