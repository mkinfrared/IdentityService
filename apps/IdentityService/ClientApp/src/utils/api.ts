import axios from "axios";

import { BASE_URL } from "utils/secrets";

const { isAxiosError } = axios;

const api = axios.create({
  baseURL: BASE_URL,
  withCredentials: true,
});

export { isAxiosError };

export default api;
