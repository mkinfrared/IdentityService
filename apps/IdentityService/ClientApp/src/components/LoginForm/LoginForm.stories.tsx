import { Meta, Story } from "@storybook/react";

import { withRouter } from "utils/storybookDecorators";

import { LoginForm } from "./LoginForm";
import { LoginFormProps } from "./LoginForm.type";
import css from "./Story.module.scss";

export default {
  title: "COMPONENTS/LoginForm",
  component: LoginForm,
  parameters: {
    componentSubtitle: "Subtitle goes here",
  },
  decorators: [withRouter],
} as Meta;

const Template: Story<LoginFormProps> = (args) => (
  <div className={css.Story}>
    <LoginForm {...args} />
  </div>
);

const Default = Template.bind({});

Default.args = {
  returnUrl: "https://localhost",
};

Default.parameters = {
  docs: {
    storyDescription: "Story description",
  },
};

export { Default };
