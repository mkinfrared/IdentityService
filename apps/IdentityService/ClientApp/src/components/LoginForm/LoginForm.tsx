import { yupResolver } from "@hookform/resolvers/yup";
import {
  classNames,
  getFormErrors,
  usePasswordVisibility,
} from "@identity-service/core";
import { Button, Heading, Text } from "@identity-service/ui";
import { AxiosError } from "axios";
import { MouseEventHandler, memo } from "react";
import { useForm } from "react-hook-form";

import { loginMutation } from "api/mutations";
import { LoginCommand } from "api/types";
import InputField from "components/InputField";
import { Routes } from "pages/Main/Main.type";
import { isAxiosError } from "utils/api";
import { loginSchema } from "utils/validationSchemas";

import css from "./LoginForm.module.scss";
import { LoginFormProps } from "./LoginForm.type";

const LoginForm = ({ className, returnUrl }: LoginFormProps) => {
  const defaultValues: LoginCommand = {
    username: "",
    password: "",
    returnUrl,
  };

  const { register, formState, handleSubmit, setError } = useForm<LoginCommand>(
    {
      resolver: yupResolver(loginSchema),
      mode: "onChange",
      defaultValues,
    },
  );

  const { errors, isValid, isSubmitting } = formState;
  const { togglePasswordVisibility, Icon, fieldType } = usePasswordVisibility();

  const handleRegisterClick: MouseEventHandler = (event) => {
    event.preventDefault();

    const { location, history } = window;
    const pathname = `/account/${Routes.REGISTER}`;
    const url = new URL(location.toString());

    url.pathname = pathname;

    history.replaceState({}, "", url.toString());
  };

  const handleForgotPasswordClick: MouseEventHandler = (event) => {
    event.preventDefault();

    const { location, history } = window;
    const pathname = `/account/${Routes.FORGOT_PASSWORD}`;
    const url = new URL(location.toString());

    url.pathname = pathname;

    history.replaceState({}, "", url.toString());
  };

  const onSubmit = async (formData: LoginCommand) => {
    try {
      await loginMutation(formData);

      window.location.href = returnUrl;
    } catch (e) {
      if (isAxiosError(e)) {
        const { response } = e as AxiosError;

        const submitErrors = getFormErrors(
          response?.data as BadRequest<LoginCommand>,
        );

        submitErrors.forEach(({ name, message, type }) => {
          setError(name, { message, type });
        });
      }

      console.error(e);
    }
  };

  return (
    <div className={classNames(css.LoginForm, className)}>
      <form className={css.form} onSubmit={handleSubmit(onSubmit)}>
        <Heading className={css.heading}>Login</Heading>
        <InputField
          {...register("username")}
          className={css.input}
          label="Username"
          error={errors.username?.message}
        />
        <InputField
          {...register("password")}
          className={css.input}
          label="Password"
          suffix={Icon}
          onSuffixClick={togglePasswordVisibility}
          type={fieldType}
          error={errors.password?.message}
        />
        <input type="text" value={returnUrl} hidden name="returnUrl" readOnly />
        <Text className={classNames(css.text, css.forgotPassword)}>
          <a href="/account/forgotPassword" onClick={handleForgotPasswordClick}>
            Forgot password?
          </a>
        </Text>
        <Text className={css.text}>
          No account yet? Consider{" "}
          <a href="/account/login" onClick={handleRegisterClick}>
            register
          </a>
        </Text>
        <div className={css.buttonWrapper}>
          <Button
            disabled={!isValid || isSubmitting}
            type="submit"
            loading={isSubmitting}
          >
            Login
          </Button>
        </div>
      </form>
    </div>
  );
};

export { LoginForm };

export default memo(LoginForm);
