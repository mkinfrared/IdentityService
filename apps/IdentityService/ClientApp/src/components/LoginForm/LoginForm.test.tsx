import { fireEvent, render, waitFor } from "@testing-library/react";

import * as mutations from "api/mutations";

import { LoginForm } from "./LoginForm";

jest.mock("api/mutations");

describe("<LoginForm />", () => {
  const mutationsMock = mutations as jest.Mocked<typeof mutations>;
  const returnUrl = "?someUrl=Marklar";
  const Component = <LoginForm returnUrl={returnUrl} />;

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it("should be defined", () => {
    expect(LoginForm).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should have a disabled button if fields are empty", () => {
    const { getByTestId } = render(Component);
    const submitButton = getByTestId("Button");

    expect(submitButton).toBeDisabled();
  });

  it("should have an enabled button if fields are filled and valid", async () => {
    const { getByTestId, container } = render(Component);
    const userNameField = container.querySelector('input[name="username"]');
    const passwordField = container.querySelector('input[name="password"]');
    const submitButton = getByTestId("Button");

    expect(submitButton).toBeDisabled();

    fireEvent.change(userNameField!, { target: { value: "foobar" } });

    fireEvent.change(passwordField!, { target: { value: "marklar" } });

    await waitFor(() => expect(submitButton).toBeEnabled());
  });

  it("should call loginMutation on submit", async () => {
    mutationsMock.loginMutation.mockResolvedValueOnce("" as any);

    const data = {
      username: "marklar",
      password: "foobar",
      returnUrl,
    };

    const { getByTestId, container } = render(Component);
    const userNameField = container.querySelector('input[name="username"]');
    const passwordField = container.querySelector('input[name="password"]');
    const submitButton = getByTestId("Button");

    fireEvent.change(userNameField!, { target: { value: data.username } });

    fireEvent.change(passwordField!, { target: { value: data.password } });

    await waitFor(() => expect(submitButton).toBeEnabled());

    await fireEvent.submit(submitButton);

    await waitFor(() => expect(mutationsMock.loginMutation).toHaveBeenCalled());

    expect(mutationsMock.loginMutation).toHaveBeenCalledTimes(1);

    expect(mutationsMock.loginMutation).toHaveBeenCalledWith(data);
  });

  it("should display an error on bad network request", async () => {
    const errorMessage = "Invalid login credentials";

    const axiosError = {
      isAxiosError: true,
      response: {
        data: {
          Username: ["Invalid login credentials"],
        },
        status: 400,
      },
    };

    mutationsMock.loginMutation.mockRejectedValueOnce(axiosError);

    const { getByTestId, getByText, container } = render(Component);
    const userNameField = container.querySelector('input[name="username"]');
    const passwordField = container.querySelector('input[name="password"]');
    const submitButton = getByTestId("Button");

    fireEvent.change(userNameField!, { target: { value: "username" } });

    fireEvent.change(passwordField!, { target: { value: "password" } });

    await waitFor(() => expect(submitButton).toBeEnabled());

    await fireEvent.submit(submitButton);

    await waitFor(() => expect(mutationsMock.loginMutation).toHaveBeenCalled());

    expect(getByText(errorMessage)).toBeInTheDocument();
  });

  it("should redirect to register page", () => {
    const replaceState = jest.spyOn(window.history, "replaceState");
    const url = "http://localhost/account/register";
    const { getByText } = render(Component);
    const link = getByText(/register/);

    fireEvent.click(link);

    expect(replaceState).toHaveBeenCalled();

    expect(replaceState).toHaveBeenCalledWith({}, "", url);
  });
});
