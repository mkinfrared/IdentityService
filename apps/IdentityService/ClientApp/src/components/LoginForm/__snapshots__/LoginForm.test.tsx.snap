// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> should match the snapshot 1`] = `
<div>
  <div
    class="LoginForm"
  >
    <form
      class="form"
    >
      <h3
        class="Heading-module_Heading__Wg8yB heading"
        data-testid="Heading"
      >
        Login
      </h3>
      <label
        class="TextField-module_TextField__QtlJl input"
        data-testid="TextField"
      >
        <h5
          class="Heading-module_Heading__Wg8yB TextField-module_label__OD4N8"
          data-testid="Heading"
        >
          Username
        </h5>
        <div
          class="TextField-module_input__SPywc"
        >
          <input
            name="username"
            type="text"
            value=""
          />
        </div>
        <p
          class="Text-module_Text__K8mLZ TextField-module_error__IGQzp"
          data-testid="Text"
        />
      </label>
      <label
        class="TextField-module_TextField__QtlJl input TextField-module_hasSuffix__-bHtc"
        data-testid="TextField"
      >
        <h5
          class="Heading-module_Heading__Wg8yB TextField-module_label__OD4N8"
          data-testid="Heading"
        >
          Password
        </h5>
        <div
          class="TextField-module_input__SPywc"
        >
          <input
            name="password"
            type="password"
            value=""
          />
          <div
            class="FakeButton-module_FakeButton__-W3X5 TextField-module_suffix__SCuij"
            data-testid="FakeButton"
            role="button"
            tabindex="0"
          >
            <svg
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
              />
            </svg>
          </div>
        </div>
        <p
          class="Text-module_Text__K8mLZ TextField-module_error__IGQzp"
          data-testid="Text"
        />
      </label>
      <input
        hidden=""
        name="returnUrl"
        readonly=""
        type="text"
        value="?someUrl=Marklar"
      />
      <p
        class="Text-module_Text__K8mLZ text"
        data-testid="Text"
      >
        No account yet? Consider
         
        <a
          href="/account/login"
        >
          register
        </a>
      </p>
      <div
        class="buttonWrapper"
      >
        <button
          class="Button-module_Button__0OJEu Button-module_opaque__O0DEF"
          data-testid="Button"
          disabled=""
          type="submit"
        >
          Login
        </button>
      </div>
    </form>
  </div>
</div>
`;
